apps/ai-image-generator/backend/src/transform-images.ts

```diff
@@ -91,32 +73,102 @@ export class ImageTransformer {
   }): Promise<Buffer> {
     const { width, height, eraseColor } = maskParams;
     const sharpImage = toSharp(this.maskImageResponse.body);
+    const desiredMaskDimensions = { ...this.toDimensions(width, height) };
+
+    // width and height are computed during the squarify step
+    sharpImage.resize({
+      fit: 'fill', // ignore the aspect ratio of the input and stretch to both provided dimensions
+      kernel: 'nearest', // use nearest neighbor interpolation to preserve hard edges of mask erase areas
+    });
+
+    this.squarifyImage(sharpImage, desiredMaskDimensions);
+    const transparentImage = await this.replaceColorWithTransparent(eraseColor, sharpImage);
+    return transparentImage.toFormat('png').toBuffer();
+  }
+
+  // take an image and make it square by adding bars to the left/right or top/bottom as needed
+  private squarifyImage(sharpImage: sharp.Sharp, newDimensions: Dimensions) {
+    const { width, height, layout } = newDimensions;
 
-    const { width: initialWidth, height: initialHeight } = await sharpImage.metadata();
-    if (width !== initialWidth || height !== initialHeight) {
-      sharpImage.resize({
-        width,
-        height,
-        fit: 'fill', // stretches to fit
+    sharpImage.resize({
+      width,
+      height,
+    });
+
+    // we only need to pad the image for portraits and landscapes -- squares are fine as is
+    if (layout === 'portrait') {
```

Comment:
Food for thought: This is an example where there's repeated code but I actually think separating it out like this and having repeated code makes it easier to read even though the extraWidth calculations are all the same.

Reply:
Yeah I agree there. I often follow a "rule of 3" in situations like this, where I just copy and paste until I find myself doing it 3 times, in which case... time for an abstraction.